### PR TITLE
update BM to use MKL BM V2 API

### PR DESCRIPTION
**Pull request by glingyan**
only compilable for MKL release after 1115

Signed-off-by: lingyan <lingyan.guo@intel.com>

### CODE REVIEW

**Comment by piiswrong** [comment]
Please ping my after you fixed the path issue and we can merge

**Comment by glingyan** [comment]
hi  Eric:  any update ?

**Comment by piiswrong** [comment]
Is path fixed here? Also could you add the old version back to webdata? Its causing 404 errors for people with old code

**Comment by glingyan** [comment]
yes，will add keep old release on from now on

发自我的 iPhone

在 2016年12月1日，上午1:22，Eric Junyuan Xie <notifications@github.com<mailto:notifications@github.com>> 写道：


Is path fixed here? Also could you add the old version back to webdata? Its causing 404 errors for people with old code

**Comment by glingyan** [comment]
1.      Path problem solved, as I said , it will rename old mkl version to mkl_<version number> dir to backup

2.      Please check new web_data pull request at

https://github.com/dmlc/web-data/pull/14

**Comment by piiswrong** [comment]
So it gets downloaded to root now? Is there any documentation that mention this change?

**Comment by glingyan** [comment]
1. Enable USE_MKL2017=1 in make/config.mk

    1.1 USE_BLAS should be atlas by default

    1.2 if you need USE_BLAS to be mkl, please navigate here to do a full MKL installation: https://registrationcenter.intel.com/en/forms/?productid=2558&licensetype=2

    1.3 By default, MKL_2017_EXPRIEMENTAL=0. If setting MKL_2017_EXPRIEMENTAL=1, MKL buffer will be created and transferred between layers to achiever much higher performance.

  2. Run 'make -jX'

    2.1 Makefile will execute "prepare_mkl.sh" to download the mkl under ~/mklml_release/mklml folder
    
        2.2.1 if previous mkl version exist, e.g. previous mkl is 20120601, ~/mklml_release/mklml will rename to ~/mklml_release/mklml.20120601
        
            2.2.1.1 if previous mkl is 20120601 and mklml.20120601 is exist, this is unlikely to happen ,default behaviour is remove mklml to avoid compiling error

    2.2 if the download failed because of proxy setting, please download it manually before make

    2.2.1 wget https://github.com/dmlc/web-data/raw/master/mxnet/mklml-release/mklml_lnx_<MKL VERSION>.tgz

    2.2.2 tar zxvf mklml_lnx_<MKL VERSION>.tgz
    
    2.2.3 mv mklml_lnx_<MKL VERSION> ~/mklml_release/mklml
    
       2.2.3.1 mkdir mklml_release if not exists
       
  3. Navigate into the python directory

**Comment by glingyan** [comment]
will checkin soon

**Comment by glingyan** [comment]
put mklroot in config.mk is a good idea, will check then

**Comment by piiswrong** [comment]
Also pooling still fails. Could you fallback to mxnet implementation for pooling until its fixed for mkl?

**Comment by piiswrong** [comment]
@glingyan Also keep the tar.gz under mxnet, extract it to mklroot and remove the superfluous directory hierachies. So MKLROOT=~ to lead to things extracted to ~/include/mkl_xxx.h and ~/lib/libmklxxx.so

**Comment by piiswrong** [comment]
Please simplify the intructions in MKL_README and put it in config.mk under the relevent variables

**Comment by glingyan** [comment]
是不是将MKL_README 放在make 目录,和config.mk同目录?

**Comment by glingyan** [comment]
MKLML_ROOT feature done
MKL_README.mk updated and move to make folder , the same folder with config.mk

File: make/config.mk [hunk 1]
```diff
@@ -59,6 +59,9 @@ USE_OPENCV = 1
 # use openmp for parallelization
```
Comment (piiswrong) [inline make/config.mk]: Sorry for not being clear. I didn't mean to move MKL_README to make/. I meant picking out the relevant sentences and put them here. 
The MKLML_ROOT, USE_MKL2016 and UES_MKL_EXPERIMENTAL options should have detailed explanation of what they do. Or at least point the user to MKL_README for more details

File: prepare_mkl.sh [hunk 2]
```diff
@@ -62,26 +62,36 @@ echo $VERSION_LINE  # Return Version Line
...
+    VERSION_LINE=`GetVersionName $HOME_MKL`
+    #echo $VERSION_LINE
     if [ $VERSION_LINE -lt $VERSION_MATCH ] ; then
       #...If it is not then downloaded and unpacked
-      wget --no-check-certificate -P $DST $MKLURL -O $DST/$ARCHIVE_BASENAME
-      tar -xzf $DST/$ARCHIVE_BASENAME -C $DST
+      wget --no-check-certificate -P $MXNET_ROOT $MKLURL -O $MXNET_ROOT/$ARCHIVE_BASENAME
+      tar -xzf $MXNET_ROOT/$ARCHIVE_BASENAME -C $MXNET_ROOT
+      #echo $HOME_MKL
+      yes | cp -rf $MXNET_ROOT/$MKL_CONTENT_DIR/* $HOME_MKL
```
Comment (piiswrong) [inline prepare_mkl.sh]: Have you confirmed this works? Do you need sudo cp?

Comment (glingyan) [inline prepare_mkl.sh]: I tested:
1. root user, put it in /usr/local
2. normal user, put it in ~
but for sudo , it is hard to do , becuase makefile need to get output from prepare_mkl.sh
if need sudo to input password, it will not work 
do you have seem any solution for this?

**Comment by glingyan** [comment]
Support Pooling Asymmetric padding filled by zeros , it will pass the test when cudnn use CUDNN_POOLING_AVERAGE_COUNT_EXCLUDE_PADDING mode

File: src/operator/pooling-inl.h [hunk 3]
```diff
@@ -50,13 +48,6 @@ struct PoolingParam : public dmlc::Parameter<PoolingParam> {
     .add_enum("sum", pool_enum::kSumPooling)
     .describe("Pooling type to be applied.");
 
-    DMLC_DECLARE_FIELD(pooling_convention).set_default(pool_enum::kValid)
```
Comment (piiswrong) [inline src/operator/pooling-inl.h]: why did you remove this?

Comment (glingyan) [inline src/operator/pooling-inl.h]: this is include by MKL2017 implemented layers integration to improve IA perf. on mxnet (#3581)
at that time , mkl pooling do not support Asymmetric padding filled by zeros, by default pool_enum::kValid will use mxnet pooling implmenetation

**Comment by glingyan** [comment]
New patch upload,  add pooling_convention again

**Comment by piiswrong** [comment]
Please rebase, make the comment in config.mk more descriptive (atleast "Please refer to MKL_README.md for details) and we can merge

**Comment by glingyan** [comment]
done, please check https://github.com/dmlc/mxnet/pull/4128

**Comment by futurely** [comment]
This PR should be closed since https://github.com/dmlc/mxnet/pull/4128 has been merged.

**Comment by glingyan** [comment]
How to close it if the original repo is deleted?
